internal/upstreamldap/upstreamldap.go

```diff
@@ -601,3 +682,62 @@ func (p *Provider) traceAuthSuccess(t *trace.Trace) {
 		trace.Field{Key: "authenticated", Value: true},
 	)
 }
+
+func (p *Provider) traceSearchBaseDiscoveryFailure(t *trace.Trace, err error) {
+	t.Step("search base discovery failed",
+		trace.Field{Key: "reason", Value: err.Error()})
+}
+
+func MicrosoftUUIDFromBinary(attributeName string) func(entry *ldap.Entry) (string, error) {
+	// validation has already been done so we can just get the attribute...
+	return func(entry *ldap.Entry) (string, error) {
+		binaryUUID := entry.GetRawAttributeValue(attributeName)
+		return microsoftUUIDFromBinary(binaryUUID)
+	}
+}
+
+func microsoftUUIDFromBinary(binaryUUID []byte) (string, error) {
+	uuidVal, err := uuid.FromBytes(binaryUUID) // start out with the RFC4122 version
+	if err != nil {
+		return "", err
+	}
+	// then swap it because AD stores the first 3 fields little-endian rather than the expected
+	// big-endian.
+	uuidVal[0], uuidVal[1], uuidVal[2], uuidVal[3] = uuidVal[3], uuidVal[2], uuidVal[1], uuidVal[0]
+	uuidVal[4], uuidVal[5] = uuidVal[5], uuidVal[4]
+	uuidVal[6], uuidVal[7] = uuidVal[7], uuidVal[6]
+	return uuidVal.String(), nil
+}
+
+func GroupSAMAccountNameWithDomainSuffix(entry *ldap.Entry) (string, error) {
+	sAMAccountNameAttribute := "sAMAccountName"
+	sAMAccountNameAttributeValues := entry.GetAttributeValues(sAMAccountNameAttribute)
+
+	if len(sAMAccountNameAttributeValues) != 1 {
+		return "", fmt.Errorf(`found %d values for attribute "%s", but expected 1 result`,
+			len(sAMAccountNameAttributeValues), sAMAccountNameAttribute,
+		)
+	}
+
+	sAMAccountName := sAMAccountNameAttributeValues[0]
+	if len(sAMAccountName) == 0 {
+		return "", fmt.Errorf(`found empty value for attribute "%s", but expected value to be non-empty`,
+			sAMAccountNameAttribute,
+		)
+	}
+
+	distinguishedName := entry.DN
+	domain, err := getDomainFromDistinguishedName(distinguishedName)
+	if err != nil {
+		return "", err
+	}
+	return sAMAccountName + "@" + domain, nil
+}
+
+func getDomainFromDistinguishedName(distinguishedName string) (string, error) {
+	domainComponents := regexp.MustCompile(",DC=|,dc=").Split(distinguishedName, -1)
```

Comment:
Since the regexp is a constant, maybe we can compile the regexp once (somewhere outside this function) instead of on every call to the function. It may improve performance a tiny bit.

Reply:
Resolved in 05afae60c2d3dd95e805ea4a2899513493a1e31d